### PR TITLE
Test Framework for Enso

### LINKED ISSUE #502: Implement an Enso-Native Test Framework
### Summary
In order to best test the standard library, we should be able to write tests in Enso as soon as possible. This task deals with creating a test harness in Enso itself that ensures that we can write tests without having to rely on host language libraries. 

### Value
We can test Enso code inside Enso, rather than having to rely on external (Java or Scala) test frameworks.

### Specification

- [x] Implement a simple BDD-style testing utility.
- [x] Ensure that it outputs readable errors in the case where tests fail, and useful logs in the case of success. 

### Acceptance Criteria & Test Cases

- The above specification has been implemented.
- The test framework is tested in itself(?)

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
Introduces an Enso-native testing framework, for future stdlib efforts (and other libraries!)
Also introduces a bunch of helper functions and autogenerated atom getters.
Fixes a bug in demand analysis, thank you @iamrecursion 

### Important Notes
Closes #502 

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/DemandAnalysis.scala [hunk 1]
```diff
@@ -152,6 +152,7 @@ case object DemandAnalysis extends IRPass {
     val usesLazyTerm = isUsageOfSuspendedTerm(name)
 
     if (isInsideCallArgument) {
+      // isInsideCallArgument shouldn't be set here for inside a block
```
Comment (radeusgd) [inline engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/DemandAnalysis.scala]: If this is unexpected, shouldn't an exception be thrown to mark that?

File: engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DemandAnalysisTest.scala [hunk 2]
```diff
@@ -9,6 +9,10 @@ import org.enso.compiler.pass.{IRPass, PassConfiguration, PassManager}
 import org.enso.compiler.test.CompilerTest
 import org.enso.interpreter.runtime.scope.LocalScope
 
+import scala.annotation.nowarn
+
+// TODO Remove this
```
Comment (radeusgd) [inline engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DemandAnalysisTest.scala]: I understand that TODOs should be marked with initials of the assigned person.

**Review by radeusgd — CHANGES_REQUESTED** [review]
Looks great, just some minor comments.

File: engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/DemandAnalysis.scala (hunk 1 above)
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/DemandAnalysis.scala]:
```suggestion
```

File: engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DemandAnalysisTest.scala [hunk 3]
```diff
@@ -9,6 +9,10 @@ import org.enso.compiler.pass.{IRPass, PassConfiguration, PassManager}
...
+@nowarn("cat=unused")
```
Comment (iamrecursion) [inline engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DemandAnalysisTest.scala]:
```suggestion
```

File: test/Test/src/List_Spec.enso [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        prod = l.fold 1 (*)
+        sum.should_equal 6
+        prod.should_equal 6
+    it "should allow checking if an element satisfies a predicate with .any" <|
+        any_even = l.any (x -> x % 2 == 0)
+        any_eq_five = l.any (== 5)
+        any_even.should_be_true
+        any_eq_five.should_be_false
+
+
```
Comment (iamrecursion) [inline test/Test/src/List_Spec.enso]:
```suggestion
```

**Review by iamrecursion — APPROVED** [review]

**Review by radeusgd — APPROVED** [review]

**Pull request merged**